Account/FxAPushMessageHandler.swift

```diff
@@ -0,0 +1,220 @@
+        }
+        return result
+    }
+}
+
+extension FxAPushMessageHandler {
+    func handleVerification() -> Success {
+        guard let account = profile.getAccount(), account.actionNeeded == .needsVerification else {
+            log.info("Account verified by server either doesn't exist or doesn't need verifying")
+            return deferMaybe(PushMessageError.noActionNeeded)
+        }
+
+        // Now we're verified, we can start syncing.
+        return account.advance().bind { _ in
+            return succeed()
+        }
+    }
+}
+
+/// An extension to handle each of the messages.
+extension FxAPushMessageHandler {
+    func handleDeviceConnected(_ data: JSON?) -> Success {
+        guard let deviceName = data?["deviceName"].string else {
+            return messageIncomplete(.deviceConnected)
+        }
+        return unimplemented(.deviceConnected, with: deviceName)
+    }
+}
+
+extension FxAPushMessageHandler {
+    func handleDeviceDisconnected(_ data: JSON?) -> Success {
+        guard let deviceID = data?["id"].string else {
+            return messageIncomplete(.deviceDisconnected)
+        }
+        return unimplemented(.deviceDisconnected, with: deviceID)
+    }
+}
+
+extension FxAPushMessageHandler {
+    func handleProfileUpdated() -> Success {
+        return unimplemented(.profileUpdated)
+    }
+}
+
+extension FxAPushMessageHandler {
+    func handlePasswordChanged() -> Success {
+        return unimplemented(.passwordChanged)
+    }
+}
+
+extension FxAPushMessageHandler {
+    func handlePasswordReset() -> Success {
+        return unimplemented(.passwordReset)
+    }
+}
+
+extension FxAPushMessageHandler {
+    func handleCollectionChanged(_ data: JSON?) -> Success {
+        guard let collections = data?["collections"].arrayObject as? [String] else {
+            log.warning("collections_changed received but incomplete: \(data)")
+            return deferMaybe(PushMessageError.messageIncomplete)
+        }
+        let sm = profile.syncManager
+        let deferreds = collections.flatMap { (id: String) -> SyncResult? in
+            switch id {
+            case "addons":
+                return nil
+            case "forms":
+                return nil
+            case "prefs":
+                return nil
+            case "bookmarks":
+                return nil
+            case "history":
+                return sm.syncHistory()
+            case "tabs":
+                return sm.syncClientsThenTabs()
+            case "passwords":
+                return sm.syncLogins()
+            case "clients":
+                // if we get "tabs", "clients" then we'll do "tabs" first, 
+                // then "client", but not "tabs" a second time.
+                return sm.syncClients()
+            default:
+                return nil
+            }
+        }
+
+        if deferreds.isEmpty {
+            return deferMaybe(PushMessageError.noActionNeeded)
+        }
+
+        return all(deferreds).bind { _ in
```

Comment:
I think it makes most sense to take the set of collections that changed as a single lump and hand it to the `SyncManager` to decide what to do — after all, that's exactly what it's for.

I'd probably generalize `syncEverything`, rather than expose `syncSeveral` as-is — `sE` returns a coarser result type, and the caller needn't concern themselves with finding a `SyncFunction`. We also want to make sure that `clients` is synced before `tabs`, but only synced once.

So we'd probably have something like

```
public func syncNamedCollections(why: SyncReason, names: [String]) -> Success {
    return self.syncSeveral(why: why, synchronizers: 
    names.flatMap { switch $0 {
        case "addons": nil,
        case "clients": ("clients", self.syncClientsWithDelegate),
        …
    }) >>> succeed
}
```

with a little extra fiddling in there to do clients first.